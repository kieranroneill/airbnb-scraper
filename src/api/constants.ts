export const BASE_ENDPOINT: string = '/api';
export const LISTING_ENDPOINT: string = '/listing';
