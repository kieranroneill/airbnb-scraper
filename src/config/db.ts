import { ConnectionOptions } from 'mongoose';

export const defaultConnectionOptions: ConnectionOptions = {
  useCreateIndex: true,
  useNewUrlParser: true,
  useUnifiedTopology: true,
};
