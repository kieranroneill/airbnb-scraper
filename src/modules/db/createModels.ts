import { Connection } from 'mongoose';

// Models.
import { IModel } from '../../models';
import { IListingModel } from '../../models/listing';

// Schemas
import { listingSchema } from '../../schemas/listing';

/**
 * Convenience function to create the db model.
 * @param {Connection} connection a mongoose connection.
 * @returns {IModel} an initialised model.
 */
export default function (connection: Connection): IModel {
  return {
    listing: connection.model<IListingModel>('Listing', listingSchema),
  };
}

