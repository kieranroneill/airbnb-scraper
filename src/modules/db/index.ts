export { default as connect } from './connect';
export { default as createModels } from './createModels';
