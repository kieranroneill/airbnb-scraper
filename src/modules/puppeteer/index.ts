export { default as getListingId } from './getListingId';
export { default as mapListingFromAirbnb } from './mapListingFromAirbnb';
export { default as scrapeListing } from './scrapeListing';
