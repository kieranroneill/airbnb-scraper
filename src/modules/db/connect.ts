import { Connection, createConnection } from 'mongoose';

// Config.
import * as dbConfig from '../../config/db';

/**
 * Convenience function for connecting to the MongoDB.
 * @param {string} uri the MongoDB URI.
 * @returns {Connection} a mongoose connection.
 */
export default function (uri: string): Connection {
  return createConnection(uri, dbConfig.defaultConnectionOptions);
}
