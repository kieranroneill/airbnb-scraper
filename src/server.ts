import { json, urlencoded } from 'body-parser';
import * as cookieParser from 'cookie-parser';
import { config } from 'dotenv';
import * as Express from 'express';
import * as Helmet from 'helmet';
import { Connection } from 'mongoose';
import * as logger from 'morgan';
import { Logger } from 'winston';

// APIs.
import { ListingAPI } from './api';

// Constants.
import * as ApiConstants from './api/constants';

// Middlewares.
import errorHandler from './middlewares/errorHandler';

// Models
import { IModel } from './models';

// Modules.
import { connect, createModels } from './modules/db';
import { createLogger } from './modules/logger';

export class Server {
  public app: Express.Application;
  public logger: Logger;

  private model!: IModel;

  constructor() {
    this.app = Express();
    this.logger = createLogger();

    this.config();
    this.api();

    // Error handling.
    this.app.use(errorHandler);
  }

  /**
   * Creates the API endpoints.
   */
  private api(): void {
    const router: Express.Router = Express.Router();

    ListingAPI.create(router, this.model);

    this.app.use(ApiConstants.BASE_ENDPOINT, router);
  }

  /**
   * Configures the application.
   */
  private config(): void {
    const mongoDbUri: string = process.env.MONGODB_URI || 'mongodb://localhost:27017/airbnb-listings';
    let connection: Connection;

    // Source env file.
    config();

    // Setup some middlewares.
    this.app.use(logger('dev'));
    this.app.use(Helmet());
    this.app.use(json());
    this.app.use(urlencoded({
      extended: true,
    }));
    this.app.use(cookieParser(process.env.COOKIE_SECRET));

    // Connect to DB.
    connection = connect(mongoDbUri);

    connection.on('connected', () => this.logger.info(`Connected to: ${mongoDbUri}`));

    // Create models.
    this.model = createModels(connection);
  }
}
