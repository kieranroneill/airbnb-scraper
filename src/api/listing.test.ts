import { Application } from 'express';
import { MongoMemoryServer } from 'mongodb-memory-server';
import { Connection, createConnection } from 'mongoose';
import { agent, Response } from 'supertest';

// Config.
import * as dbConfig from '../config/db';

// Enums.
import * as ApiConstants from './constants';

// Modules.
import * as db from '../modules/db';

// Server.
import { Server } from '../server';

jest.mock('../modules/db/connect');

const endpoint: string = `${ApiConstants.BASE_ENDPOINT}${ApiConstants.LISTING_ENDPOINT}`;

interface IScope {
  app: Application;
}

describe(`${endpoint}`, () => {
  let mongod: MongoMemoryServer;
  let scope: IScope;

  beforeAll(async () => {
    let connection: Connection;
    let uri: string;

    mongod = new MongoMemoryServer();
    uri = await mongod.getConnectionString();
    connection = createConnection(uri, dbConfig.defaultConnectionOptions);

    // Mock the wrapped connection.
    (db.connect as jest.Mock).mockReturnValue(connection);
  });

  afterAll(async () => {
    await mongod.stop();
  });

  beforeEach(() => {
    scope = {
      app: new Server().app,
    };
  });

  describe(`POST ${ApiConstants.LISTING_ENDPOINT}`, () => {
    it('should return a 400 if no url is provided', async () => {
      const response: Response = await agent(scope.app)
        .post(`${endpoint}`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    it('should return a 400 if the url is not an Airbnb url listing', async () => {
      const response: Response = await agent(scope.app)
        .post(`${endpoint}`)
        .send({
          url: 'http://somewhere.someplace.not.nice',
        })
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });

    it('should return a 400 if the url has no listing id', async () => {
      const response: Response = await agent(scope.app)
        .post(`${endpoint}`)
        .send({
          url: 'https://airbnb.co.uk/rooms/',
        })
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });
});
