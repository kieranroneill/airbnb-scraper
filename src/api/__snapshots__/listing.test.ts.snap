// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/listing POST /listing should return a 400 if no url is provided 1`] = `
Object {
  "message": "This is not an Airbnb URL",
  "status": 400,
}
`;

exports[`/api/listing POST /listing should return a 400 if the url has no listing id 1`] = `
Object {
  "message": "No listing id found in URL",
  "status": 400,
}
`;

exports[`/api/listing POST /listing should return a 400 if the url is not an Airbnb url listing 1`] = `
Object {
  "message": "This is not an Airbnb URL",
  "status": 400,
}
`;
